File: algorithm/LeetCode/AddBinary/AddBinary.cpp
class Solution {
public:
    string addBinary(string a, string b) {
        reverse(a.begin(), a.end());
        reverse(b.begin(), b.end());
        int len1 = a.length();
        int len2 = b.length();
        string result;
        int flag ,val;
        flag = val = 0;
        int i;
        for(i = 0; i < len1 && i < len2; ++i){
            val = (a[i] - '0') + (b[i] - '0') + flag;
            result.append(1, val % 2 + '0');
            flag = val/2;
        }
        while (i < len1) {
            val = a[i] - '0'+ flag;
            result.append(1, val % 2 + '0');
            flag = val/2;
            ++i;
        }
        while (i < len2 ){
            val = b[i] - '0'+ flag;
            result.append(1, val % 2 + '0');
            flag = val/2;
            ++i;
        }
        cout << "flag " << flag << endl;
        if(flag != 0)
            result.append(1, '1');
        reverse(result.begin(), result.end());
        return result;
    }
};

File: algorithm/LeetCode/AddBinary/AddBinary.java
public class Solution {
    public String addBinary(String a, String b) {
        int len1 = a.length();
        int len2 = b.length();
        String ret = new String();
        int i = 0;
        int carry = 0;
        a = ReverseStr(a);
        b = ReverseStr(b);
        while(i < len1 || i < len2 || carry != 0){
            int val = ((i<len1)?(a.charAt(i)-'0'):0) + ((i<len2)?(b.charAt(i)-'0'):0) + carry;
            carry = 0;
            if(val > 1){
                carry = val/2;
                ret += (char)((val%2) + '0');
            }else{
                ret += (char)(val + '0');
            }
            i++;
        }
        return ReverseStr(ret);
    }

       public String ReverseStr(String str){
        return new StringBuffer(str).reverse().toString();
    }
}